docs/source/rfc/isa_restructuring.org

```diff
@@ -0,0 +1,732 @@
+to achieve with the current architecture, but it would result in many
+redundant pieces of code.
+
+There is also the matter of mixing levels of abstraction. =Mint= is a
+financial operation, =If= is logical, =FailBox=, probably represents
+an instruction that fails always, but why does it have the word =Box=
+in it? More importantly by being specific we lose flexibility to allow
+different kinds of blockchains to cohesively specialise. It is poorly
+defined whether a =Transfer= would allow exchange of currencies in the
+blockchain, and with instructions being defined host-side and having
+no other mechanism of expressing the concepts, we are in the
+precarious position of not being able to leave this up to the
+individual blockchains. Further, the entire set of =Grant= and
+=Revoke= instructions is made useless in the public use case, nudging
+developers of public blockchains to use this functionality as it
+happened with =Metadata=.
+
+In addition to conceptual redundancy, there is also redundancy between
+different elements of the ISA. We have both a =SequenceBox=, and
+multiple instructions comprising a single transaction. There's both an
+=If= as an =instruction= and an =If= as an =Expression=. Further
+obfuscating the types, each instruction accepts an expression which
+either wraps a =Value= or a collection of operations that are
+ill-defined for some of the potential variants of =Value=.
+
+Finally, compounding the issue, the Iroha special instructions were
+meant to be accompanied by a domain-specific language that is only now
+being written.
+
+* Proposed ISA
+
+There are multiple ways to take advantage of the Executor RFC that
+primarily exploit one property. The API of Iroha is part dynamically
+loadable logic that can be upgraded, and provides natural versioning
+and part strict frozen set of basic operations that are verbose, but
+guarantee forwards and backwards compatibility.
+
+** Variant 0: Executor-defined Expressions and Instructions
+
+The idea is that the Instructions and the Expressions are both
+implementation details of the Executor. This has the following
+advantages:
+- Simple to implement[fn:2]
+- WSV owned by Executor and can be treated like a black box
+- Executor upgrades replace versioning, all transactions are stored
+  with no extra information.
+- If libraries are statically linked, produces the largest throughput.
+- The instruction set can be completely altered and tailored to
+  specific blockchain needs. Expressions can be eliminated or extended
+  to arbitrarily complex operations.
+- All logic is defined in one place and thus easy to audit.
+- All logic can be patched at runtime.
+
+This approach necessarily requires the Executor to be implemented in a
+compiled language. Further, while it can cope with the current choice
+of WASM, it is affected by some of the limitations of said binary
+format. For instance: because WASM is 32-bit-only, only 4GiB of world
```

Comment:
What about [wasm64-unknown-unknown](https://doc.rust-lang.org/rustc/platform-support/wasm64-unknown-unknown.html)?

Reply:
As per [rust-lang](https://doc.rust-lang.org/rustc/platform-support/wasm64-unknown-unknown.html):

> This target is not a stable target. The [memory64](https://github.com/webassembly/memory64) WebAssembly proposal is still in-progress and not standardized. This means that there are not many engines which implement the memory64 feature and if they do they're likely behind a flag, for example: ...

Which I should include in the documentation.